### PR TITLE
Validation layer to the API

### PR DESCRIPTION
**Pull request by owl352**
# Issue
Currently, request are not really validated, there are several specific checks implemented in the controllers, but we should have a way to set a specific requests data format for the each API endpoint.

Fastify is being used as an http server in the api, and it is already shipped with the support of the JSON Schema, which ideally fits our needs

# Things done
Now we can validate requests via fastify by requsts schemas.
The code from the controllers that validate of the entered parameters was removed

### CODE REVIEW

File: packages/api/test/integration/data.contracts.spec.js [hunk 1]
```diff
@@ -290,21 +290,8 @@ describe('DataContracts routes', () => {
       assert.deepEqual(body, expectedDataContract)
     })
 
-    //
```
Comment (pshenmic) [inline packages/api/test/integration/data.contracts.spec.js:293]: Don't delete this test, I think I forgot to reimplement that during one of the refactors

File: packages/api/test/integration/data.contracts.spec.js [hunk 2]
```diff
@@ -290,21 +290,8 @@ describe('DataContracts routes', () => {
...
-    //     assert.equal(body.identifier, 'Gc7HqRGqmA4ZSafQ6zXeKH8Rh4AjNjjWsztotJDLpMXa')
-    //     assert.equal(body.txHash, '4107CE20DB3BE2B2A3B3F3ABA9F68438428E734E4ACF39D4F6D03B0F9B187829')
-    //     assert.equal(body.owner, 'FRMXvU2vRqk9xTya3MTB58ieBt27izpPyoX3fVLf3HuA')
-    //     assert.equal(body.version, 2)
-    //     assert.equal(body.timestamp, '2024-02-22T14:23:57.592Z')
-    // });
-
     it('should return 404 if data contract not found', async () => {
-      await client.get('/dataContract/DEADBEEFDEADBEEFDEADBEEFDEADBEEFDEADBEEFDEADBEEFDEADBEEFDEADBEEF')
+      await client.get('/dataContract/Gc70000000000afQ6zXeKH8Rh4AjNjjWsztotJDLpMXa')
```
Comment (pshenmic) [inline packages/api/test/integration/data.contracts.spec.js:294]: Just use something from the production environment (testnet)

File: packages/api/test/integration/identities.spec.js [hunk 3]
```diff
@@ -68,7 +68,7 @@ describe('Identities routes', () => {
     })
 
     it('should return 404 when identity not found', async () => {
-      await client.get('/identity/fake')
+      await client.get('/identity/12345678901234567890123456789012345678901234')
```
Comment (pshenmic) [inline packages/api/test/integration/identities.spec.js:71]: Just use something from the production environment (testnet)

File: packages/api/src/schemas.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  },
+  validatorsSchema: {
+    querystring: { $ref: 'paginationWithIsActiveSchema#' }
+  },
+  validatorSchema: {
+    params: { $ref: 'txHashSchema#' }
+  }
+}
+
+// function initSchemas(fastify) {
```
Comment (pshenmic) [inline packages/api/src/schemas.js:152]: Comment

File: packages/api/src/schemas.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
+const schemaObjects = [
+  {
+    $id: 'txHashSchema',
+    type: 'object',
+    properties: {
+      txHash: { type: 'string', minLength: 64, maxLength: 64 }
+    }
+  }, {
```
Comment (pshenmic) [inline packages/api/src/schemas.js:8]: Make object on each line

File: packages/api/src/controllers/ValidatorsController.js [hunk 6]
```diff
@@ -26,23 +26,13 @@ class ValidatorsController {
...
-        return response.status(400).send({ message: `invalid isActive value ${order}. only boolean values are accepted` })
-      }
-    }
-
     const validators = await this.validatorsDAO.getValidators(
       Number(page),
       Number(limit),
       order,
-      typeof isActive === 'undefined' ? undefined : isActive === 'true',
+      isActive,
```
Comment (pshenmic) [inline packages/api/src/controllers/ValidatorsController.js:35]: Will it work for undefined (when query var isActive not set)?

**Review by pshenmic — CHANGES_REQUESTED** [review]

File: packages/api/src/schemas.js (hunk 4 above)
Comment (owl352) [reply packages/api/src/schemas.js:161]: ✅

File: packages/api/src/controllers/ValidatorsController.js (hunk 6 above)
Comment (owl352) [reply packages/api/src/controllers/ValidatorsController.js:35]: ✅

File: packages/api/src/schemas.js (hunk 5 above)
Comment (owl352) [reply packages/api/src/schemas.js:17]: ✅

File: packages/api/test/integration/data.contracts.spec.js (hunk 1 above)
Comment (owl352) [reply packages/api/test/integration/data.contracts.spec.js:293]: ✅

File: packages/api/test/integration/data.contracts.spec.js (hunk 2 above)
Comment (owl352) [reply packages/api/test/integration/data.contracts.spec.js]: ✅

File: packages/api/src/schemas.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+      isActive: { type: ['boolean', 'null'] },
+      page: { $ref: 'baseSchema#/properties/page' },
+      limit: { $ref: 'baseSchema#/properties/limit' },
+      order: { $ref: 'baseSchema#/properties/order' }
+    }
+  }
+]
+
+const schemas = {
+  epochSchema: { params: { $ref: 'epochSchema#' } },
```
[the remark this answers is not in the document]
Comment (owl352) [reply packages/api/src/schemas.js:87]: Done

File: packages/api/src/schemas.js [hunk 8]
```diff
@@ -0,0 +1,128 @@
+const schemaObjects = [
+  {
+    $id: 'baseSchema',
+    type: 'object',
+    properties: {
+      hash: { type: 'string', minLength: 64, maxLength: 64 },
+      page: { type: ['integer', 'null'], minimum: 1 },
+      limit: { type: ['integer', 'null'], minimum: 1, maximum: 100 },
```
[the remark this answers is not in the document]
Comment (owl352) [reply packages/api/src/schemas.js:8]: Done

File: packages/api/src/schemas.js [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+      order: { type: ['string', 'null'], enum: ['asc', 'desc'] }
+    }
+  },
+  {
+    $id: 'txHashSchema',
+    type: 'object',
+    properties: { txHash: { $ref: 'baseSchema#/properties/hash' } }
```
[the remark this answers is not in the document]
Comment (owl352) [reply packages/api/src/schemas.js:15]: Done

File: packages/api/src/routes.js [hunk 10]
```diff
@@ -27,109 +27,145 @@ module.exports = ({
...
-      handler: transactionsController.getTransactions
+      handler: transactionsController.getTransactions,
+      schema: { querystring: { $ref: 'paginationSchema#' } }
     },
     {
       path: '/transaction/:txHash',
       method: 'GET',
-      handler: transactionsController.getTransactionByHash
+      handler: transactionsController.getTransactionByHash,
+      schema: { params: { $ref: 'txHashSchema#' } }
```
Comment (pshenmic) [inline packages/api/src/routes.js:61]: make params on the next line

File: packages/api/src/schemas.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    type: 'object',
+    properties: { identifier: { type: 'string', minLength: 43, maxLength: 44 } }
+  },
+  {
+    $id: 'epochSchema',
+    type: 'object',
+    properties: { index: { type: 'number', minimum: 0 } }
+  },
+  {
+    $id: 'paginationWithOrderBySchema',
```
Comment (pshenmic) [inline packages/api/src/schemas.js:63]: Why 3 different paginations?

File: packages/api/src/schemas.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    type: 'object',
+    properties: {
+      page: { $ref: 'page#' },
+      limit: { $ref: 'limit#' },
+      order: { $ref: 'order#' },
+      orderBy: { $ref: 'orderBy#' }
+    }
+  },
+  {
+    $id: 'paginationSchema',
```
Comment (pshenmic) [inline packages/api/src/schemas.js:73]: This should be a type instead

File: packages/api/src/schemas.js [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    type: 'object',
+    properties: { txHash: { $ref: 'hash#' } }
+  },
+  {
+    $id: 'validatorSchema',
+    type: 'object',
+    properties: { validator: { $ref: 'hash#' } }
+  },
+  {
+    $id: 'hashSchema',
```
Comment (pshenmic) [inline packages/api/src/schemas.js:48]: No need a separate object

**Review by pshenmic — CHANGES_REQUESTED** [review]

**Review by pshenmic — APPROVED** [review]

**Review by pshenmic — APPROVED** [review]
Lgtm

**Pull request merged**